initialization/initialization.go

```diff
@@ -170,17 +213,68 @@ func (init *Initializer) Initialize(ctx context.Context) error {
 
 	numLabels := uint64(init.opts.NumUnits) * uint64(init.cfg.LabelsPerUnit)
 	fileNumLabels := numLabels / uint64(init.opts.NumFiles)
+	difficulty := shared.PowDifficulty(numLabels)
+	batchSize := uint64(config.DefaultComputeBatchSize)
 
 	init.logger.Info("initialization: starting to write %v file(s); number of units: %v, number of labels per unit: %v, number of bits per label: %v, datadir: %v",
 		init.opts.NumFiles, init.opts.NumUnits, init.cfg.LabelsPerUnit, init.cfg.BitsPerLabel, init.opts.DataDir)
 
 	for i := 0; i < int(init.opts.NumFiles); i++ {
-		if err := init.initFile(ctx, uint(init.opts.ComputeProviderID), i, numLabels, fileNumLabels); err != nil {
+		if err := init.initFile(ctx, i, batchSize, numLabels, fileNumLabels, difficulty); err != nil {
 			return err
 		}
 	}
 
-	return nil
+	if init.nonce != nil {
+		return nil
+	}
+
+	init.logger.Info("initialization: no nonce found while computing leaves, continue searching")
+	if init.lastPosition == nil || *init.lastPosition < numLabels {
+		init.lastPosition = new(uint64)
+		*init.lastPosition = numLabels
+	}
+
+	// continue searching for a nonce
+	// TODO(mafa): PM-195 depending on the difficulty function this can take a VERY long time, with the current difficulty function
+	// ~ 37% of all smeshers won't find a nonce while computing leaves
+	// ~  2% of all smeshers won't find a nonce even after checking 4x numLabels
+	for i := *init.lastPosition; i < math.MaxUint64; i += batchSize {
+		defer init.saveMetadata()
```

Comment:
It doesn't seem that you are updating `lastPosition` according to the latest PoW-only batch. So upon recovery, you will repeat the same unsuccessful positions.

Reply:
Updated accordingly